File: rust/tests/adc_dac/mcp4725_test/src/main.rs
use linux_embedded_hal::I2cdev;
use periph::chips::adc_dac::Mcp4725Full;

macro_rules! check_true {
    ($cond:expr, $label:expr, $passed:expr, $failed:expr) => {
        if $cond {
            println!("PASS {}", $label);
            $passed += 1;
        } else {
            println!("FAIL {}", $label);
            $failed += 1;
        }
    };
}

fn main() {
    let i2c_bus: u8 = std::env::var("I2C_BUS").ok().and_then(|v| v.parse().ok()).unwrap_or(1);
    let addr: u8 = std::env::var("I2C_ADDR")
        .ok()
        .and_then(|v| u8::from_str_radix(v.trim_start_matches("0x"), 16).ok())
        .unwrap_or(0x60);

    let dev = I2cdev::new(format!("/dev/i2c-{}", i2c_bus)).expect("open i2c bus");
    let mut dac = Mcp4725Full::new(dev, addr).expect("init MCP4725");

    let mut passed = 0i32;
    let mut failed = 0i32;

    dac.set_voltage(0.5).unwrap();
    check_true!(true, "set_voltage_0.5", passed, failed);

    dac.set_raw(2048).unwrap();
    check_true!(true, "set_raw_2048", passed, failed);

    dac.set_voltage_eeprom(0.5).unwrap();
    check_true!(true, "set_voltage_eeprom_0.5", passed, failed);

    dac.set_raw_eeprom(2048).unwrap();
    check_true!(true, "set_raw_eeprom_2048", passed, failed);

    let (code, vf, pd, ec, epd, ready) = dac.read().unwrap();
    check_true!(code <= 4095, "read_code_range", passed, failed);
    check_true!(vf >= 0.0 && vf <= 1.0, "read_voltage_fraction_range", passed, failed);
    check_true!(ec <= 4095, "read_eeprom_code_range", passed, failed);

    dac.set_power_down(0).unwrap();
    check_true!(true, "set_power_down_normal", passed, failed);

    dac.set_power_down(1).unwrap();
    check_true!(true, "set_power_down_1k", passed, failed);

    dac.set_power_down(2).unwrap();
    check_true!(true, "set_power_down_100k", passed, failed);

    dac.set_power_down(3).unwrap();
    check_true!(true, "set_power_down_500k", passed, failed);

    dac.wake_up().unwrap();
    check_true!(true, "wake_up", passed, failed);

    dac.reset().unwrap();
    check_true!(true, "reset", passed, failed);

    let ready = dac.is_eeprom_ready().unwrap();
    check_true!(ready == true || ready == false, "is_eeprom_ready_bool", passed, failed);

    println!("===DONE: {} passed, {} failed===", passed, failed);
    std::process::exit(if failed == 0 { 0 } else { 1 });
}

File: go/tests/adc_dac/mcp4725_test/main.go
//go:build linux && !tinygo

// MCP4725 hardware test — Linux host.
//
// Drives every method in the MCP4725Full API and reports PASS/FAIL per
// check. Prints the standard ===DONE: ... === summary line at the end.
// Exits 0 on full pass, 1 on any failure.
package main

import (
	"fmt"
	"os"
	"strconv"

	"github.com/tuhde/Periph/go/periph/chips/adc_dac"
	"github.com/tuhde/Periph/go/periph/transport"
)

func main() {
	bus, err := strconv.Atoi(envOr("I2C_BUS", "1"))
	if err != nil {
		fmt.Fprintln(os.Stderr, "I2C_BUS:", err)
		os.Exit(2)
	}
	addr, err := strconv.ParseUint(envOr("I2C_ADDR", "0x60"), 0, 8)
	if err != nil {
		fmt.Fprintln(os.Stderr, "I2C_ADDR:", err)
		os.Exit(2)
	}

	tr, err := transport.NewI2CTransport(bus, uint8(addr))
	if err != nil {
		fmt.Fprintln(os.Stderr, "transport:", err)
		os.Exit(2)
	}
	defer tr.Close()

	chip, err := adcdac.NewMCP4725Full(tr)
	if err != nil {
		fmt.Fprintln(os.Stderr, "new:", err)
		os.Exit(2)
	}

	passed, failed := 0, 0
	check := func(label string, cond bool) {
		if cond {
			fmt.Println("PASS", label)
			passed++
		} else {
			fmt.Println("FAIL", label)
			failed++
		}
	}

	if err := chip.SetVoltage(0.5); err != nil {
		fmt.Fprintln(os.Stderr, "set_voltage_0.5:", err)
	}
	check("set_voltage_0.5", true)

	if err := chip.SetRaw(2048); err != nil {
		fmt.Fprintln(os.Stderr, "set_raw_2048:", err)
	}
	check("set_raw_2048", true)

	if err := chip.SetVoltageEEPROM(0.5); err != nil {
		fmt.Fprintln(os.Stderr, "set_voltage_eeprom_0.5:", err)
	}
	check("set_voltage_eeprom_0.5", true)

	if err := chip.SetRawEEPROM(2048); err != nil {
		fmt.Fprintln(os.Stderr, "set_raw_eeprom_2048:", err)
	}
	check("set_raw_eeprom_2048", true)

	st, err := chip.Read()
	if err != nil {
		fmt.Fprintln(os.Stderr, "read:", err)
	}
	check("read_code_range", st.Code <= 4095)
	check("read_voltage_fraction_range", st.VoltageFraction >= 0.0 && st.VoltageFraction <= 1.0)
	check("read_eeprom_code_range", st.EEPROMCode <= 4095)

	if err := chip.SetPowerDown(0); err != nil {
		fmt.Fprintln(os.Stderr, "set_power_down_normal:", err)
	}
	check("set_power_down_normal", true)

	if err := chip.SetPowerDown(1); err != nil {
		fmt.Fprintln(os.Stderr, "set_power_down_1k:", err)
	}
	check("set_power_down_1k", true)

	if err := chip.SetPowerDown(2); err != nil {
		fmt.Fprintln(os.Stderr, "set_power_down_100k:", err)
	}
	check("set_power_down_100k", true)

	if err := chip.SetPowerDown(3); err != nil {
		fmt.Fprintln(os.Stderr, "set_power_down_500k:", err)
	}
	check("set_power_down_500k", true)

	if err := chip.WakeUp(); err != nil {
		fmt.Fprintln(os.Stderr, "wake_up:", err)
	}
	check("wake_up", true)

	if err := chip.Reset(); err != nil {
		fmt.Fprintln(os.Stderr, "reset:", err)
	}
	check("reset", true)

	ready, err := chip.IsEEPROMReady()
	if err != nil {
		fmt.Fprintln(os.Stderr, "is_eeprom_ready:", err)
	}
	check("is_eeprom_ready_bool", ready == true || ready == false)

	fmt.Printf("===DONE: %d passed, %d failed===\n", passed, failed)
	if failed != 0 {
		os.Exit(1)
	}
}

func envOr(k, def string) string {
	if v, ok := os.LookupEnv(k); ok {
		return v
	}
	return def
}
